openedx/core/djangoapps/persistenttask/task.py

```diff
@@ -0,0 +1,44 @@
+"""
+Celery utility code for persistent tasks.
+"""
+
+from datetime import datetime
+
+from celery import Task
+import pytz
+
+from .models import FailedTask
+
+
+# pylint: disable=abstract-method
+class PersistOnFailureTask(Task):
+    """
+    Custom Celery Task base class that persists task data on failure.
+    """
+
+    def on_failure(self, exc, task_id, args, kwargs, einfo):
+        """
+        If the task fails, persist a record of the task.
+        """
+        def truncate_to_field(model, field_name, value):
+            """
+            If data is too big for the field, it would cause a failure to
+            insert, so we shorten it, truncating in the middle (because
+            valuable information often shows up at the end.
+            """
+            field = model._meta.get_field(field_name)  # pylint: disable=protected-access
+            if len(value) > field.max_length:
+                first = value[:(field.max_length // 2)]
+                last = value[(field.max_length // 2) + 3:field.max_length]
```

Comment:
It looks like this calculation needs to be corrected.  This doesn't extract the end of the string as expected.  Since it extracts only until the `max_length` index of the string.